### ISSUE TITLE
Proposal: explicit definition of a scene/depicted_world coordinate system

### ISSUE THREAD

**Issue by alisterburt** [napari/napari#3848]
## 🚀 Feature
@kevinyamauchi and I have been discussing a potential move away from the specific 0,1,2 nomenclature for the 'displayed/depicted world' towards an explicit definition a scene coordinate system. Layer dims would then be mapped into this coordinate system in the same way as they are now.

This is helpful because certain non-data objects don't really exist in the nD world, only in the depicted world.

## Motivation

When data has mixed dimensionality, the axis visual cannot correctly map to dimensions on all layers. 
![image](https://user-images.githubusercontent.com/7307488/146637255-02eae76a-e3ce-4d5e-9467-a43c6c4cace2.png)

This can be confusing, it happens because we follow numpy broadcasting rules which start by aligning the dimensions to the right. See some zulip discussion [here](https://napari.zulipchat.com/#narrow/stream/212875-general/topic/issues.20with.20axes/near/232199053)

We currently have two features which interact with the 'scene coordinate system' in 3D, plane rendering (#3023 / #3759) and clipping planes #3252 and @kevinyamauchi and I have been working on a couple more to enable general 3D interactivity. 
Key observation:
- these objects only exist in 3D, not nD

Being able to specifically place/orient non-data objects within the 2d/3d scene and not worry about the whole nD world would simplify the addition of controls around these things. In #3759 there is discussion about what to label the orientation controls which I've currently labeled z/y/x and bound to those keys. Without definition of a 'scene coordinate system', these keybindings would have to change depending on which dimensions were currently depicted

## Pitch

- explicitly document and refer to a 'scene coordinate system', this could be `z/y/x`
- add some sort of visual indicator showing which of the data dims are mapped to `z/y/x`

The visual indicator could be
- a widget with a simple table like the one above
- a faint set of 2/3 numbers next to the layer name in the layer list, showing which 2/3 dims are being displayed

## Alternatives

- accept the current inconsistency
- name dimensions -1/-2/-3 

Naming dimensions -1, -2, -3 is consistent in nD but confusing to think about - it also doesn't solve the problem of having consistent labels/keybindings for orienting things in the scene.

## Additional context

Certain properties like extents also only make sense when discussed in the context of what is actually depicted, what is actually depicted would only live in this `depicted_world` coordinate system. This came up recently in #2104 and also for extents on the image layer where we had an extra half pixel depicted on each side of the world extents.

cc @kevinyamauchi @jni

**Comment by jni** [comment]
Thanks @alisterburt!

> When data has mixed dimensionality, the axis visual cannot correctly map to dimensions on all layers.

I've thought about this a lot (indeed this inconsistency has haunted my dreams since the very early days of napari!) and have come to think that napari (and NumPy) in fact align the dimensions of the *broadcasted* arrays, ie the result of `np.broadcast_to`. In other words, by the time napari is displaying things, there is *no* mixed dimensionality: all arrays have been broadcast to have the same dimensions.

Currently, napari falls completely short of this, though (@perlman and I ran into this experimenting with slicing): if you have arrays A of shape (128, 256, 512) and B of (256, 512), displaying them together in 3D works as expected, but setting `dims.order` to (2, 0, 1) shows the 256 axes of A and B aligned, but the 512 axis of A gets aligned with the 128 axis of B, which is rubbish, incorrect in any universe, and a consequence of having implicit rather than explicit coordinate models.

```python
import numpy as np
import napari

A = np.random.random((128, 256, 512))
B = np.random.random((256, 512))

viewer = napari.Viewer()
viewer.add_image(A, colormap='magenta', blending='additive')
viewer.add_image(B, colormap='green', blending='additive')
viewer.dims.order = (2, 0, 1)

napari.run()
```

In other words, the mapping of dimensions between different arrays changes depending on the dims being displayed!!! :exploding_head: Seriously I find this so embarrassing. :joy: 

So, the first fix I would do is change this particular behaviour and make sure that axes are indeed broadcast as NumPy would do — and named/referred to appropriately. This might actually involve some pre/post-broadcast axis number wrangling at various parts of the code base, which is pretty unfortunate. :stuck_out_tongue_winking_eye: 

But @alisterburt 

> add some sort of visual indicator showing which of the data dims are mapped to z/y/x

This is super genius, and not only helps clarify napari's handling of dimensions: it also helps me clarify in my head why I was [so opposed](https://github.com/ome/ngff/issues/35#issuecomment-787899798) to naming data array axes in the OME-NGFF spec discussion, though I was (temporarily, I hope) overruled. :sweat_smile: 

But yeah we need to be clear on what the mapping shows. We could do one mapping per layer, then we can show the "original" axes mapped to xyz, or we do a single mapping, in which case we need to show the *broadcast* axes, and this means that we need some documentation around how napari handles mixed dimensionality.

Anyway, love this. :blush:

**Comment by kevinyamauchi** [comment]
> Currently, napari falls completely short of this, though (@perlman and I ran into this experimenting with slicing): if you have arrays A of shape (128, 256, 512) and B of (256, 512), displaying them together in 3D works as expected, but setting dims.order to (2, 0, 1) shows the 256 axes of A and B aligned, but the 512 axis of A gets aligned with the 128 axis of B, which is rubbish, incorrect in any universe, and a consequence of having implicit rather than explicit coordinate models.

I agree the current behavior is not intuitive and probably incorrect. @alisterburt and I have been fighting this a lot in our 3D interactivity work. In case anybody passing by is wondering how the the aligning dimensions of data with mixed dimensionality currently works, see [here](https://github.com/napari/napari/blob/942b68809250973fcc1fd7a6dfbc38914f29b0ef/napari/layers/utils/layer_utils.py#L592-L632).

> add some sort of visual indicator showing which of the data dims are mapped to z/y/x

I think this is a good first step that should probably go in soon-ish. While our current broadcasting behavior is undesirable, at least having an explicit display of the mapping would allow us to start referring to the "displayed dimensions" (or whatever we want to call them), which as @alisterburt said above, massively simplifies the 3D interactivity UI.

**Comment by JoOkuma** [comment]
I really like the idea.
I think it would also be beneficial to store and indicate the 3 spatial (scene) dimensions even when only a 2D slice is being shown and not only the displayed dimensions. Because a plugin or a layer might want to have a different interaction when scrolling through a spatial axis or a temporal (or whatever it would be called) axes.

**Comment by alisterburt** [comment]
Ooph - thanks for the example @jni, I hadn't realised the mapping changed depending on the specific dims and their order, unfortunate is the right word! 

> But yeah we need to be clear on what the mapping shows. We could do one mapping per layer, then we can show the "original" axes mapped to xyz, or we do a single mapping, in which case we need to show the broadcast axes, and this means that we need some documentation around how napari handles mixed dimensionality.

totally agree - my gut feeling (and I think what I would expect as a user) would be to go with showing how the original axes are mapped, happy to be overruled here though

> how the the aligning dimensions of data with mixed dimensionality currently works, see here.

With what @kevinyamauchi has done here it seems like we can tackle this work in two parts
1. implement a corresponding inverse `Layer._data_to_world` transform for the layer dims and add a visual indicator for the current system (however inconsistent)
2. work to make the system consistent as @jni described independent of the visual component

One thing we haven't tackled is the naming of the axes - naming of the axes is one of the pieces blocking #3759, if we could come to some agreement here on the eventual naming of the axes in the scene coordinate system that might unblock things there

My vote for that is `x/y/z` which map to array dimensions `-1/-2/-3` without any reordering/mixed dimensionality shenanigans

**Comment by sofroniewn** [comment]
Seems like everyone here is in-agreement that things should change!

> In other words, the mapping of dimensions between different arrays changes depending on the dims being displayed!!! 🤯 Seriously I find this so embarrassing. 😂

In the example above - should we just treat every 2D array as actually being 3D with a dimension of length 1 at the beginning? During 3D slicing?

```python
B = np.random.random((1, 256, 512))
```

Would that fix everything?

2D slicing should always work fine, and 3D slicing with arrays >= 3D should work fine. The only issue is when we have 3D slicing with an array of dim 2 right?

This should be easy to fix, and I'd just consider a bug.

Unless I'm missing something much deeper here

**Comment by sofroniewn** [comment]
Or do we also get problems when say mixing a 4D array and a 5D array?

Either way, I believe the lines of code that will need to be changed are here https://github.com/napari/napari/blob/9849b5c77d9dfc0e46c5c24012b865fee51109d1/napari/layers/base/base.py#L951-L962

where we go from a global dims model to the "layer" dims model (which can have lower dimensionality)

**Comment by brisvag** [comment]
Love this conversation, and agree with most things said above!

If we extend the coordinate space nomenclature, I would steer away from `scene`, since that is basically the equivalent of `world` in vispy terms, and it would be confusing. I would lean towards `depicted`, `displayed`, or something along those lines.

I'm also gonna be *that guy* and bring up `xarray` again, just to remind everyone about the possibility of having named dimensions take care of this :P

In the meantime, showing the mapping is a great compromise!

**Comment by alisterburt** [comment]
I did some experiments with the code you linked @sofroniewn to understand in which situations we have a problem
<details>
<summary>example code</summary>

```python
import numpy as np


def world_order_to_layer_order(order, global_ndim, layer_ndim):
    # adjust the order of the global dims based on the number of
    # dimensions that a layer has - for example a global order of
    # [2, 1, 0, 3] -> [0, 1] for a layer that only has two dimensions
    # or -> [1, 0, 2] for a layer with three as that corresponds to
    # the relative order of the last two and three dimensions
    # respectively
    offset = global_ndim - layer_ndim
    order = np.array(order)
    if offset <= 0:
        order = list(range(-offset)) + list(order - offset)
    else:
        order = list(order[order >= offset] - offset)
    return order

print('------ 3d world, 2d vs 3d layer ------')
world_order = [(0, 1, 2), (2, 0, 1)]
world_ndim = 3
layer_ndim = [2, 3]

for order_ in world_order:
    for layer_ndim_ in layer_ndim:
        layer_order = world_order_to_layer_order(order_, world_ndim, layer_ndim_)
        print(f"       world ndim: {world_ndim}\n"
              f"       layer ndim: {layer_ndim_}\n"
              f" world dims order: {order_}\n"
              f" layer dims order: {tuple(layer_order)}\n")


print('------ 5d world, 3d vs 4d vs 5d layer ------')
world_order = [(0, 1, 2, 3, 4), (2, 0, 1, 3, 4)]
world_ndim = 5
layer_ndim = [3, 4, 5]

for order_ in world_order:
    for layer_ndim_ in layer_ndim:
        layer_order = world_order_to_layer_order(order_, world_ndim, layer_ndim_)
        print(f"       world ndim: {world_ndim}\n"
              f"       layer ndim: {layer_ndim_}\n"
              f" world dims order: {order_}\n"
              f" layer dims order: {tuple(layer_order)}\n")
```

</details>

<details>
<summary>results</summary>

```text
------ 3d world, 2d vs 3d layer ------
       world ndim: 3
       layer ndim: 2
 world dims order: (0, 1, 2)
 layer dims order: (0, 1)

       world ndim: 3
       layer ndim: 3
 world dims order: (0, 1, 2)
 layer dims order: (0, 1, 2)

       world ndim: 3
       layer ndim: 2
 world dims order: (2, 0, 1)
 layer dims order: (1, 0)

       world ndim: 3
       layer ndim: 3
 world dims order: (2, 0, 1)
 layer dims order: (2, 0, 1)

------ 5d world, 3d vs 4d vs 5d layer ------
       world ndim: 5
       layer ndim: 3
 world dims order: (0, 1, 2, 3, 4)
 layer dims order: (0, 1, 2)

       world ndim: 5
       layer ndim: 4
 world dims order: (0, 1, 2, 3, 4)
 layer dims order: (0, 1, 2, 3)

       world ndim: 5
       layer ndim: 5
 world dims order: (0, 1, 2, 3, 4)
 layer dims order: (0, 1, 2, 3, 4)

       world ndim: 5
       layer ndim: 3
 world dims order: (2, 0, 1, 3, 4)
 layer dims order: (0, 1, 2)

       world ndim: 5
       layer ndim: 4
 world dims order: (2, 0, 1, 3, 4)
 layer dims order: (1, 0, 2, 3)

       world ndim: 5
       layer ndim: 5
 world dims order: (2, 0, 1, 3, 4)
 layer dims order: (2, 0, 1, 3, 4)
```

</details>
So I think we also have a problem when 5d is sliced down to 2/3/4d and you're right, shouldn't be too crazy to fix!

---- 

> I would steer away from scene, since that is basically the equivalent of world in vispy terms, and it would be confusing

I quite like `scene` - to explain a `depicted_world` to someone you have to start by explaining an `nD_world`, to explain `scene` you just say *'oh, it's the thing you're currently looking at!'*. 
I don't disagree that this could be confusing but actually the fact that we have our nD world and vispy has its 3D world is already pretty confusing and seems worse to me... generally I think we often mirror vispy but we're actually an abstraction over it and can be a bit more separate where it makes sense (with good docs for those who delve into the vispy world)

> I'm also gonna be that guy and bring up xarray again, just to remind everyone about the possibility of having named dimensions take care of this :P

😝 *down the rabbit hole we go*
Isn't it more confusing/error prone to describe mappings between multiple systems with similar nomenclature? This is imo the main reason to choose `z/y/x` as names which array dimension indices are mapped into and I *think* what @jni was getting at with 

> it also helps me clarify in my head why I was so opposed to naming data array axes in the OME-NGFF spec discussion

**Comment by kevinyamauchi** [comment]
>> I'm also gonna be that guy and bring up xarray again, just to remind everyone about the possibility of having named dimensions take care of this :P
>>
> 😝 down the rabbit hole we go
> Isn't it more confusing/error prone to describe mappings between multiple systems with similar nomenclature? This is imo the main reason to choose z/y/x as names which array dimension indices are mapped into and I think what @jni was getting at with

I think there's been a lot of great discussion here! I'd like to clarify the current consensus just to make sure I am on the same page.

The summary below is my understanding of the current consensus. Do you all agree?

### depicted coordinate system

The first step is to expand our defined coordinate systems to include a `depicted` coordinate system that is the coordinate system of the scene being displayed in the canvas. The axes of the `depicted` coordinate system have static and defined names. This is distinct from naming the layer data axes, which can be named as the user wishes (i.e., napari does not specify layer data axis names). `dims_displayed` (could be renamed) then maps the `depicted` axes to layer data axes. @alisterburt suggested the `depicted` coordinate system should have names such as (`z`, `y`, `x`) for 3D and (`y`, `x`) for 2D. I like this idea, as these names can't be confused with dimension indices. 

Do we want to store the mapping of layer data axes to `depicted` axes? Something like this?
```python
# dimensions without names
layer._dims_displayed
>>> {'z': 0, 'y': 1, 'x': 2}

# dimensions with names
layer._dims_displayed
>>> {'z': 'time', 'y': 'y_c', 'x': 'x_c'}
```
**A few questions:**
1. Are we all on board with calling the "displayed coordinate system" axes `z`, `y`, and `x`? I think we unblock @alisterburt on #3759 if we agree on the axis names.
2. Where do we want to store the mapping between the `depicted` axes and layer data axes? On the layer? I know we are trying to move dims stuff off of the layers, but `dims_displayed` is currently on the layer.
3. Do we want to call the "displayed coordinate system" `depicted`? I agree with @alisterburt that `scene` might be a good alternative. In the context of napari, I disagree that the vispy scene coordinate system is similar to the napari world coordinate system. The napari world coordinate system is nD, while the vispy scene coordinate system is just the 2 or 3 dims being displayed (although vispy is xyz and napari is zyx). Thus, I think calling our "displayed coordinate system" scene is fairly apt.

### depicted coordinate system visual

The next step is to create visual representation of the current mapping of layer data axes to `depicted` coordinate system axes. Based on @alisterburt 's schematic above, it would be something like:

```
depicted: [0, 1, 2] 
 layer_0: [0, 1, 2]
 layer_1:    [0, 1]
```

### fix broadcasting of layer data

Currently, the way napari aligns dims for layers with mixed dimensionality is counter-intuitive. It should be modified so that layers data are expanded to match the dimensionality of the world coordinate system. For example, if the world coordinate system is 5D and one of the layers is 2D with shape `(y, x)`, it would be treated as if it's shape where `(1, 1, 1, y, x)`

**Comment by brisvag** [comment]
> > Isn't it more confusing/error prone to describe mappings between multiple systems with similar nomenclature? This is imo the main reason to choose z/y/x as names which array dimension indices are mapped into and I think what @jni was getting at with

> Do we want to store the mapping of layer data axes to `depicted` axes? Something like this?
> 
> ```python
> # dimensions without names
> layer._dims_displayed
> >>> {'z': 0, 'y': 1, 'x': 2}
> 
> # dimensions with names
> layer._dims_displayed
> >>> {'z': 'time', 'y': 'y_c', 'x': 'x_c'}
> ```

I think @kevinyamauchi 's second example is the same thing as having native named dimensions, but with extra steps. The first example is almost the same, except instead of naming we force the user to decide how to map each layer to a dimension. I would say that either we do everything automatically with broadcasting (as now, but fixing the inconsistencies) so layers are unaware of what we're doing, or we might as well introduce named dimensions as first class citizens :P

>     1. Are we all on board with calling the "displayed coordinate system" axes `z`, `y`, and `x`? I think we unblock @alisterburt on [volume plane interactive controls #3759](https://github.com/napari/napari/pull/3759) if we agree on the axis names.

Works for me!

>     2. Where do we want to store the mapping between the `depicted` axes and layer data axes? On the layer? I know we are trying to move dims stuff off of the layers, but `dims_displayed` is currently on the layer.

See above.

>  The napari world coordinate system is nD, while the vispy scene coordinate system is just the 2 or 3 dims being displayed (although vispy is xyz and napari is zyx). Thus, I think calling our "displayed coordinate system" scene is fairly apt.

Yeah I don't know what I was thinking, you're totally right. Disregard my comment about this :)

**Comment by kevinyamauchi** [comment]
>I think @kevinyamauchi 's second example is the same thing as having native named dimensions, but with extra steps. The first example is almost the same, except instead of naming we force the user to decide how to map each layer to a dimension. I would say that either we do everything automatically with broadcasting (as now, but fixing the inconsistencies) so layers are unaware of what we're doing, or we might as well introduce named dimensions as first class citizens :P

Sorry, I think my explanation was a bit unclear. I did not mean to suggest that the user provides the mapping between `depicted` axes and `layer.data` axes. I think the mapping should be done automatically, as it is now (but with broadcasting fixed). I meant to ask if we want to create a new property or update the existing `layer._dims_displayed` to contain an explicit mapping between `depicted` axes and the `layer.data` axes rather than the current list of indices.

Regarding named `layer.data` dimensions, I think that is a parallel, but separate issue. I am definitely open to adding named `layer.data` dimensions in the future as long as the user can choose them as they wish (within reason). I don't think napari should specify layer data dimension names.

**Comment by kevinyamauchi** [comment]
> Yeah I don't know what I was thinking, you're totally right. Disregard my comment about this :)

No worries! So many coordinate systems! Once we decide on this `depicted` / `scene` / `displayed` coordinate system, I'll work on an explanatory doc. I think that will be super useful both for us making sure we are all on the same page and for users.

**Comment by sofroniewn** [comment]
> Are we all on board with calling the "displayed coordinate system" axes z, y, and x? I think we unblock @alisterburt on volume plane interactive controls #3759 if we agree on the axis names.

I'm fine with this too.

> Where do we want to store the mapping between the depicted axes and layer data axes? On the layer? I know we are trying to move dims stuff off of the layers, but dims_displayed is currently on the layer.

I'm fine with this as long as it stays private for now. Especially if mapping is done automatically. Once we fix slicing we can move off layer

> Do we want to call the "displayed coordinate system" depicted? I agree with @alisterburt that scene might be a good alternative.

I don't mind. I think this will also be private for now, so we can change

> It should be modified so that layers data are expanded to match the dimensionality of the world coordinate system. For example, if the world coordinate system is 5D and one of the layers is 2D with shape (y, x), it would be treated as if it's shape where (1, 1, 1, y, x)

Yes - let's do that. I think that could be a standalone PR where we just add a new `world_order_to_layer_order` function like what @alisterburt has done in https://github.com/napari/napari/issues/3848#issuecomment-1000511719, but with the new behavior fixed. Ideally I think we'd make that PR first as I think we can fix that, get that merged and then do the new stuff

Do folks now feel unblocked on this, know how to move forward?

**Comment by kevinyamauchi** [comment]
>>  Do we want to call the "displayed coordinate system" depicted? I agree with @alisterburt that scene might be a good alternative.
>>
> I don't mind. I think this will also be private for now, so we can change

I am now leaning towards `scene` for the name of the "displayed coordinate system". I think it's the most descriptive and concise name. I agree it should be private for now.

>> It should be modified so that layers data are expanded to match the dimensionality of the world coordinate system. For example, if the world coordinate system is 5D and one of the layers is 2D with shape (y, x), it would be treated as if it's shape where (1, 1, 1, y, x)
>>
> Yes - let's do that. I think that could be a standalone PR where we just add a new world_order_to_layer_order function like what @alisterburt has done in #3848 (comment), but with the new behavior fixed. Ideally I think we'd make that PR first as I think we can fix that, get that merged and then do the new stuff

A standalone PR makes sense to me and I agree it can go first. I think @alisterburt is unblocked on #3759 as long as we are all in agreement that `z`, `y`, and `x` are the `scene` axis names (and it sounds like that's the case), so the other parts are not as urgent.

> Do folks now feel unblocked on this, know how to move forward?

I think it's all good from my side. Thoughts @alisterburt, @brisvag, and @jni (and any others)?

**Comment by alisterburt** [comment]
Thanks for all of the input here all! I definitely feel unblocked, that we have a way forward and that the discussion has been super constructive 🚀 

@kevinyamauchi `scene` is my favorite too and starting with this hidden away in private attributes seems like the way to go :)

**Comment by jni** [comment]
Just a small note that "scene" has previously been described as "worldslice" or "sliced world" coordinates, but maybe scene is better. I'm not sure. But either way I wanted to make sure everyone knows that correspondence because it helps to search for relevant issues and PRs. :blush:

**Comment by brossetti** [comment]
I see that `_dims_displayed`, `_dims_displayed_order`, `_dims_not_displayed`, and `_dims_order` have been added as private attributes. This is extremely helpful for me because I recently discovered this problem while working on my own plugin (many thanks to @brisvag for leading me here). Is there any chance these attributes can be made public? They are needed for my plugin, and (according to the FutureWarning) I won't be able to use them in v0.5.0.
